src/bin/ipa_bench/gen_events.rs

```diff
@@ -203,109 +79,71 @@ pub fn generate_events<R: RngCore + CryptoRng, W: io::Write>(
     }
 }
 
-fn gen_events<R: RngCore + CryptoRng>(
-    params: &GenEventParams,
-    secret_share: bool,
+fn gen_reports<R: RngCore + CryptoRng>(
+    impressions: u8,
+    conversions: u8,
+    epoch: &<EventTimestamp as EpochDuration>::Epoch,
+    breakdown_key: Number,
     sample: &Sample,
     rng: &mut R,
-    ss_rng: &mut R,
-) -> Vec<Event> {
-    let mut events: Vec<Event> = Vec::new();
-
-    let matchkeys = gen_matchkeys(params.matchkeys, rng);
-    let mut ss_mks: Vec<SecretShare> = Vec::new();
-
-    if secret_share {
-        for mk in &matchkeys {
-            // Currently, all geneerated match keys are set in all source events from the same user. This is an ideal
-            // scenario where all devices are used equally. In reality, however, that isn't the case. Should we pick
-            // a few match keys out from the events?
-            ss_mks.push(mk.xor_split(ss_rng));
-        }
-    }
+) -> Vec<GenericReport> {
+    let mut reports: Vec<GenericReport> = Vec::new();
 
-    // Randomly choose a datetime of the first impression in [0..DAYS_IN_EPOCH)
-    // TODO: Assume that impressions happen any time within the epoch
-    let mut last_impression = Duration::new(rng.gen_range(0..DAYS_IN_EPOCH * 24 * 60 * 60), 0);
-
-    for _ in 0..params.impressions {
-        let t = last_impression + sample.impressions_time_diff(rng);
-
-        if secret_share {
-            events.push(Event::EncryptedSource(ESourceEvent {
-                event: EEvent {
-                    matchkeys: ss_mks.clone(),
-                    //TODO: Carry to next epoch if timestamp > DAYS_IN_EPOCH
-                    epoch: params.epoch,
-                    timestamp: u32::try_from(t.as_secs()).unwrap().xor_split(ss_rng),
-                },
-                breakdown_key: params.breakdown_key,
-            }));
-        } else {
-            events.push(Event::Source(SourceEvent::new(
-                matchkeys.clone(),
-                //TODO: Carry to next epoch if timestamp > DAYS_IN_EPOCH
-                params.epoch,
-                u32::try_from(t.as_secs()).unwrap(),
-                params.breakdown_key,
-            )));
-        }
+    let matchkey = gen_matchkey(rng);
+    let base_epoch: EventTimestamp = (u64::from(*epoch) * EventTimestamp::SECONDS_IN_EPOCH).into();
```

Comment:
as @akoshelev had mentioned in a diff PR, it would be nice to wrap this timestamp arithmetic in a struct to hide all these kinds of conversions

Reply:
I'll create `add_epoch()`.

Reply:
Thanks! I created `From<Epoch>'.